src/gil.rs

```diff
@@ -165,6 +175,15 @@ impl GILGuard {
 /// The Drop implementation for `GILGuard` will release the GIL.
 impl Drop for GILGuard {
     fn drop(&mut self) {
+        // Safety check & update counter
+        GILGUARD_COUNT.with(|c| {
+            if c.get() - 1 != self.previous_count {
+                panic!("GILGuards must be dropped in the reverse of the order they are created.")
```

Comment:
**We should panic only when the gstate is `PyGILState_UNLOCKED` and GILGuard is nested.**
In this nested case we care, `gstate` is always `PyGILState_LOCKED` and it's not problematic to set it in an incorrect order.

Reply:
What I meant is:
```rust
GIL_COUNT.try_with(|c| {
    if c.get() > 0 and self.gstate == PyGILState_UNLOCKED { panic!("...") }
});
```

Reply:
Good idea, I'll go for that. Avoids needing to add `GILGUARD_COUNT`